core/server/master/src/main/java/alluxio/master/file/loadmanager/LoadManager.java

```diff
@@ -87,393 +83,276 @@ public final class LoadManager {
+   * Get load jobs.
+   * @return load jobs
+   */
+  @VisibleForTesting
+  public Map<String, Load> getLoadJobs() {
+    return mLoadJobs;
+  }
 
-      LoadRequest loadRequest = buildRequest(load, blockBatch);
-      ListenableFuture<LoadResponse> listenableFuture = client.load(loadRequest);
-      Futures.addCallback(listenableFuture, new FutureCallback<LoadResponse>() {
-        @Override
-        public void onSuccess(@Nonnull LoadResponse r) {
-          TaskStatus s = r.getStatus();
+  private void run() {
+    while (!Thread.currentThread().isInterrupted()) {
+      if (System.currentTimeMillis() - WORKER_UPDATE_INTERVAL > mWorkerLastUpdateTime) {
+        updateWorkers();
+      }
+      mLoadTasks.forEach(this::processJob);
+    }
+  }
 
-          /*If successful, just schedule a next run on loadBlockBatch;
-           only handles for TaskStatus not successful.*/
-          if (s != TaskStatus.SUCCESS) {
-            List<BlockStatus> blockStatusList = r.getBlockStatusList();
-            blockStatusList.forEach(blockStatus -> {
-              Block block = blockStatus.getBlock();
-              if (blockStatus.getCode() != Status.Code.OK.value()) {
-                if (blockStatus.getRetryable()) {
-                  blockBuffer.addToRetry(block);
-                } else { //if blocks failed, handle this condition
-                  blockBuffer.addFailedBlock(block);
-                  load.addFailedBlock(block.getBlockId(), blockStatus.getMessage());
-                  /*The corresponding ufs must be FAILURE status*/
-                  load.addUfsLoadStatus(block.getUfsPath(), TaskStatus.FAILURE);
+  private void processJob(Load load, Map<WorkerInfo, LoadTask> loadTasks) {
+    switch (load.getStatus()) {
+      case SUBMITTED:
+        Preconditions.checkState(
+            loadTasks.size() == 0,
+            String.format("New job should not have outstanding requests, but load %s has %d",
+                load.getPath(), mLoadTasks.get(load).size()));
+        mActiveWorkers.forEach((workerInfo, workerClient) -> {
+          Optional<LoadTask> newTask = scheduleBatch(load, workerClient, BATCH_SIZE);
+          newTask.map(task -> loadTasks.put(workerInfo, task));
+        });
+        load.setStatus(Load.LoadStatus.PROCESSING);
+        break;
+      case PROCESSING:
+        Set<WorkerInfo> loadWorkers = loadTasks.keySet();
+        Map<WorkerInfo, CloseableResource<BlockWorkerClient>> newWorkers =
+            mActiveWorkers.entrySet().stream()
+                .filter(entry -> !loadWorkers.contains(entry.getKey()))
+                .collect(ImmutableMap.toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
+        for (Iterator<Map.Entry<WorkerInfo, LoadTask>> it = loadTasks.entrySet().iterator();
+            it.hasNext();) {
+          Map.Entry<WorkerInfo, LoadTask> entry = it.next();
+          WorkerInfo workerInfo = entry.getKey();
+          LoadTask loadTask = entry.getValue();
+          if (loadTask.isDone()) {
```

Comment:
is it possible to let the task notify the load manager when complete instead of the manager checking all the time?